### ISSUE TITLE
Faster parsing by changing source location representation

### ISSUE THREAD

**Issue by expipiplus1** [haskell-nix/hnix#746]
I noticed that the documentation for megaparsec says that [`getSourcePos` "is *not cheap*, do not call it e.g. on matching of every token, that's a bad idea"](https://hackage.haskell.org/package/megaparsec-9.0.1/docs/Text-Megaparsec.html#v:getSourcePos).

We currently call this twice for every expression. I wonder if it would be possible to change the source position representation in the parser to be a token (character) offset instead, which is very cheap ([`getOffset`](https://hackage.haskell.org/package/megaparsec-9.0.1/docs/Text-Megaparsec.html#v:getOffset)). If the user has requested a parse without source positions then this is discarded, else it's transformed into full line and column representation (sufficient laziness could even make this automatic, the location translation isn't even performed until the pretty error message needs to be printed for example).

This would also benefit downstream users who actually want the character offset and who currently have to reconstruct this from the line and column position!

Defining `getSourcePos = pure (SourcePos "" (mkPos 0) (mkPos 0))` does speed things up significantly:

![image](https://user-images.githubusercontent.com/857308/97794985-b4fc1680-1c3b-11eb-841a-b317354b93d8.png)

**Comment by layus** [comment]
That's a neat improvement :-). How much work does this represent ?

**Comment by expipiplus1** [comment]
I suspect that given the very pleasing type of expressions in hnix it shouldn't be too much trouble at all.

- Very easy to switch to the character offsets, just changing a handful of lines in the parser.
- Writing a function to convert these to line/col representation, would just be a recursive traversal of the expression tree keeping state of where we are in the file.

I think writing tests would probably be the most time-consuming part :D

Not sure if there are any performance penalties lurking where I haven't thought about though.

**Comment by Anton-Latukha** [comment]
In #1026 I also note that the current `Pos` type produces in `Expr.Types` 10 orphan instances, notification on which currently suppressed.

**Comment by Anton-Latukha** [comment]
What do you think about https://hackage.haskell.org/package/monoid-subclasses-1.1.2/docs/Data-Monoid-Instances-Positioned.html.

**Comment by Anton-Latukha** [comment]
Or just use: https://hackage.haskell.org/package/megaparsec-9.0.1/docs/Text-Megaparsec.html#v:PosState

**Comment by expipiplus1** [comment]
Yeah, the latter might be good, but it's still calculating line/col offsets
strictly which might be the slow bit.
